### ISSUE TITLE
DART locality valgrind issues

### ISSUE THREAD

**Issue by devreal** [dash-project/dash#72]
Valgrind reports the following memory access issues related to DART locality when running the DASH test suite

```
==19862== Conditional jump or move depends on uninitialised value(s)
==19862==    at 0x5A7CCAE: std::ostreambuf_iterator<char, std::char_traits<char> > std::num_put<char, std::ostreambuf_iterator<char, std::char_traits<char> > >::_M_insert_int<long>(std::ostreambuf_iterator<char, std::char_traits<char> >, std::ios_base&, char, long) const (in /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.21)
==19862==    by 0x5A7CEDC: std::num_put<char, std::ostreambuf_iterator<char, std::char_traits<char> > >::do_put(std::ostreambuf_iterator<char, std::char_traits<char> >, std::ios_base&, char, long) const (in /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.21)
==19862==    by 0x5A893F9: std::ostream& std::ostream::_M_insert<long>(long) (in /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.21)
==19862==    by 0x812341: std::enable_if<std::is_integral<int>::value, dash::util::LocalityJSONPrinter&>::type dash::util::operator<< <int>(dash::util::LocalityJSONPrinter&, int const&) (LocalityJSONPrinter.h:76)
==19862==    by 0x810645: dash::util::LocalityJSONPrinter::print_domain(int, dart_domain_locality_s const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) (LocalityJSONPrinter.cc:87)
==19862==    by 0x70038B: dash::util::LocalityJSONPrinter::operator<<(dart_domain_locality_s const&) (LocalityJSONPrinter.h:41)
==19862==    by 0x6FD932: LocalRangeTest_ArrayBlockcyclic_Test::TestBody() (LocalRangeTest.cc:14)
==19862==    by 0x7F57C2: void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) (in /home/joseph/src/dash/dash/build/bin/dash-test-mpi)
==19862==    by 0x7EF71C: void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) (in /home/joseph/src/dash/dash/build/bin/dash-test-mpi)
==19862==    by 0x7D3DB3: testing::Test::Run() (in /home/joseph/src/dash/dash/build/bin/dash-test-mpi)
==19862==    by 0x7D474B: testing::TestInfo::Run() (in /home/joseph/src/dash/dash/build/bin/dash-test-mpi)
==19862==    by 0x7D4E3E: testing::TestCase::Run() (in /home/joseph/src/dash/dash/build/bin/dash-test-mpi)
==19862==
==19862== Use of uninitialised value of size 8
==19862==    at 0x5A7BB13: ??? (in /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.21)
==19862==    by 0x5A7CCD9: std::ostreambuf_iterator<char, std::char_traits<char> > std::num_put<char, std::ostreambuf_iterator<char, std::char_traits<char> > >::_M_insert_int<long>(std::ostreambuf_iterator<char, std::char_traits<char> >, std::ios_base&, char, long) const (in /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.21)
==19862==    by 0x5A7CEDC: std::num_put<char, std::ostreambuf_iterator<char, std::char_traits<char> > >::do_put(std::ostreambuf_iterator<char, std::char_traits<char> >, std::ios_base&, char, long) const (in /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.21)
==19862==    by 0x5A893F9: std::ostream& std::ostream::_M_insert<long>(long) (in /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.21)
==19862==    by 0x812341: std::enable_if<std::is_integral<int>::value, dash::util::LocalityJSONPrinter&>::type dash::util::operator<< <int>(dash::util::LocalityJSONPrinter&, int const&) (LocalityJSONPrinter.h:76)
==19862==    by 0x810645: dash::util::LocalityJSONPrinter::print_domain(int, dart_domain_locality_s const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) (LocalityJSONPrinter.cc:87)
==19862==    by 0x70038B: dash::util::LocalityJSONPrinter::operator<<(dart_domain_locality_s const&) (LocalityJSONPrinter.h:41)
==19862==    by 0x6FD932: LocalRangeTest_ArrayBlockcyclic_Test::TestBody() (LocalRangeTest.cc:14)
==19862==    by 0x7F57C2: void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) (in /home/joseph/src/dash/dash/build/bin/dash-test-mpi)
==19862==    by 0x7EF71C: void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) (in /home/joseph/src/dash/dash/build/bin/dash-test-mpi)
==19862==    by 0x7D3DB3: testing::Test::Run() (in /home/joseph/src/dash/dash/build/bin/dash-test-mpi)
==19862==    by 0x7D474B: testing::TestInfo::Run() (in /home/joseph/src/dash/dash/build/bin/dash-test-mpi)
==19862==
==19862== Conditional jump or move depends on uninitialised value(s)
==19862==    at 0x5A7BB1F: ??? (in /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.21)
==19862==    by 0x5A7CCD9: std::ostreambuf_iterator<char, std::char_traits<char> > std::num_put<char, std::ostreambuf_iterator<char, std::char_traits<char> > >::_M_insert_int<long>(std::ostreambuf_iterator<char, std::char_traits<char> >, std::ios_base&, char, long) const (in /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.21)
==19862==    by 0x5A7CEDC: std::num_put<char, std::ostreambuf_iterator<char, std::char_traits<char> > >::do_put(std::ostreambuf_iterator<char, std::char_traits<char> >, std::ios_base&, char, long) const (in /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.21)
==19862==    by 0x5A893F9: std::ostream& std::ostream::_M_insert<long>(long) (in /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.21)
==19862==    by 0x812341: std::enable_if<std::is_integral<int>::value, dash::util::LocalityJSONPrinter&>::type dash::util::operator<< <int>(dash::util::LocalityJSONPrinter&, int const&) (LocalityJSONPrinter.h:76)
==19862==    by 0x810645: dash::util::LocalityJSONPrinter::print_domain(int, dart_domain_locality_s const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) (LocalityJSONPrinter.cc:87)
==19862==    by 0x70038B: dash::util::LocalityJSONPrinter::operator<<(dart_domain_locality_s const&) (LocalityJSONPrinter.h:41)
==19862==    by 0x6FD932: LocalRangeTest_ArrayBlockcyclic_Test::TestBody() (LocalRangeTest.cc:14)
==19862==    by 0x7F57C2: void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) (in /home/joseph/src/dash/dash/build/bin/dash-test-mpi)
==19862==    by 0x7EF71C: void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) (in /home/joseph/src/dash/dash/build/bin/dash-test-mpi)
==19862==    by 0x7D3DB3: testing::Test::Run() (in /home/joseph/src/dash/dash/build/bin/dash-test-mpi)
==19862==    by 0x7D474B: testing::TestInfo::Run() (in /home/joseph/src/dash/dash/build/bin/dash-test-mpi)

==19862== Conditional jump or move depends on uninitialised value(s)
==19862==    at 0x5A7CD0C: std::ostreambuf_iterator<char, std::char_traits<char> > std::num_put<char, std::ostreambuf_iterator<char, std::char_traits<char> > >::_M_insert_int<long>(std::ostreambuf_iterator<char, std::char_traits<char> >, std::ios_base&, char, long) const (in /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.21)
==19862==    by 0x5A7CEDC: std::num_put<char, std::ostreambuf_iterator<char, std::char_traits<char> > >::do_put(std::ostreambuf_iterator<char, std::char_traits<char> >, std::ios_base&, char, long) const (in /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.21)
==19862==    by 0x5A893F9: std::ostream& std::ostream::_M_insert<long>(long) (in /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.21)
==19862==    by 0x812341: std::enable_if<std::is_integral<int>::value, dash::util::LocalityJSONPrinter&>::type dash::util::operator<< <int>(dash::util::LocalityJSONPrinter&, int const&) (LocalityJSONPrinter.h:76)
==19862==    by 0x810645: dash::util::LocalityJSONPrinter::print_domain(int, dart_domain_locality_s const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) (LocalityJSONPrinter.cc:87)
==19862==    by 0x70038B: dash::util::LocalityJSONPrinter::operator<<(dart_domain_locality_s const&) (LocalityJSONPrinter.h:41)
==19862==    by 0x6FD932: LocalRangeTest_ArrayBlockcyclic_Test::TestBody() (LocalRangeTest.cc:14)
==19862==    by 0x7F57C2: void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) (in /home/joseph/src/dash/dash/build/bin/dash-test-mpi)
==19862==    by 0x7EF71C: void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) (in /home/joseph/src/dash/dash/build/bin/dash-test-mpi)
==19862==    by 0x7D3DB3: testing::Test::Run() (in /home/joseph/src/dash/dash/build/bin/dash-test-mpi)
==19862==    by 0x7D474B: testing::TestInfo::Run() (in /home/joseph/src/dash/dash/build/bin/dash-test-mpi)
==19862==    by 0x7D4E3E: testing::TestCase::Run() (in /home/joseph/src/dash/dash/build/bin/dash-test-mpi)
==19862==
```

IMO we should aim at being valgrind-clean to prevent bugs caused by unexpected or undefined behavior.

**Comment by fuchsto** [comment]
@devreal 
I second this. Please integrate valgrind in the CircleCI build, see issue #75

**Comment by fuchsto** [comment]
@devreal Looks like a false positive. Can you provide the exact command used to call valgrind?

Tried with:

``` bash
$ mpirun -n 2 \
      valgrind --leak-check=full --show-reachable=yes --log-file=dash-test.%p.vg \
      ./bin/dash-test-mpi
```

Did not try the valgrind MPI wrappers, these would remove lots of false positives in the MPI runtime.

**Comment by devreal** [comment]
I am running locally (GCC 5.4.0, valgrind 3.11.0) using

`mpirun -n 2 valgrind  --track-origins=yes --log-file=dash-test.%p.vg build/bin/dash-test-mpi`

I tested it on our Linux cluster but do not see the errors there (GCC 6.1.0, valgrind 3.8.1). I will re-run locally with your patches once they are merged.

**Comment by fuchsto** [comment]
Thank you, my valgrind command was identical in principle, see PR #82 

Actually we should use the valgrind MPI wrappers to eliminate false positives.
See: http://valgrind.org/docs/manual/mc-manual.html#mc-manual.mpiwrap

**Comment by devreal** [comment]
I wasn't aware of the wrappers. Building them right now to do some tests

**Comment by fuchsto** [comment]
Not sure if the valgrind MPI wrappers grok MPI 3.x, though.

**Comment by fuchsto** [comment]
Valgrind MPI wrappers do not report leaks or illegal access/free, resolved in #82
